crates/miden-lib/src/note/well_known_note.rs

```diff
@@ -157,4 +164,242 @@ impl WellKnownNote {
+            },
+
+            // if the note is `SWAP`, try to execute it to determine whether it could be consumed
+            _ => Ok(None),
+        }
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+/// Reads the account ID from the first two note input values.
+///
+/// Returns None if the note input values used to construct the account ID are invalid.
+fn try_read_account_id_from_inputs(note_inputs: &[Felt]) -> Result<AccountId, StaticAnalysisError> {
+    let account_id_felts: [Felt; 2] = note_inputs[0..2].try_into().map_err(|source| {
+        StaticAnalysisError::build_error_with_source(
+            "should be able to convert the first two note inputs to an array of two Felt elements",
+            source,
+        )
+    })?;
+
+    AccountId::try_from([account_id_felts[1], account_id_felts[0]]).map_err(|source| {
+        StaticAnalysisError::build_error_with_source(
+            "failed to create an account ID from the first two note inputs",
+            source,
+        )
+    })
+}
+
+// HELPER STRUCTURES
+// ================================================================================================
+
+/// Describes if a note could be consumed under a specific conditions: target account state
+/// and block height.
+///
+/// The status does not account for any authorization that may be required to consume the
+/// note, nor does it indicate whether the account has sufficient fees to consume it.
+#[derive(Debug)]
+pub enum NoteConsumptionStatus {
+    /// The note can be consumed by the account at the specified block height.
+    Consumable,
+    /// The note can be consumed by the account after the required block height is achieved.
+    ConsumableAfter(BlockNumber),
+    /// The note can be consumed by the account if proper authorization is provided.
+    ConsumableWithAuthorization,
+    /// The note cannot be consumed by the account at the specified conditions (i.e., block
+    /// height and account state).
+    UnconsumableConditions,
+    /// The note cannot be consumed by the specified account under any conditions.
+    NeverConsumable(Box<dyn Error + Send + Sync + 'static>),
+}
+
+#[derive(thiserror::Error, Debug)]
+#[error("failed to perform note static analysis: {message}")]
+struct StaticAnalysisError {
+    /// Stack size of `Box<str>` is smaller than String.
+    message: Box<str>,
+    /// thiserror will return this when calling Error::source on StaticAnalysisError.
+    source: Option<Box<dyn Error + Send + Sync + 'static>>,
+}
+
+impl StaticAnalysisError {
+    /// Creates a static analysis error from an error message.
+    pub fn build_error(message: impl Into<String>) -> Self {
+        let message: String = message.into();
+        Self { message: message.into(), source: None }
+    }
+
+    /// Creates a static analysis error from an error message and a source error.
+    pub fn build_error_with_source(
```

Comment:
Nit: I'd call these `new` and `with_source` or something along those lines, since they are not actually "build" related errors here.